File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/dto/BookStoreViewTest.java
package org.babyfish.jimmer.sql.dto;

import org.babyfish.jimmer.sql.model.BookStore;
import org.babyfish.jimmer.sql.model.BookStoreDraft;
import org.babyfish.jimmer.sql.model.dto.BookStoreView;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.Collections;
import java.util.UUID;

import static org.babyfish.jimmer.sql.common.Tests.assertContentEquals;

public class BookStoreViewTest {
    @Test
    public void testDtoToEntity() {
        final UUID uuid = UUID.randomUUID();
        BookStoreView view = new BookStoreView();
        view.setAvgPrice(BigDecimal.ONE);
        final BookStoreView.TargetOf_newestBooks newestBooks = new BookStoreView.TargetOf_newestBooks();
        newestBooks.setName("book1");
        newestBooks.setEdition(1);
        view.setNewestBooks(Collections.singletonList(newestBooks));
        view.setNewestBookIds(Collections.singletonList(uuid));
        assertContentEquals(
                "{" +
                        "--->\"avgPrice\":1," +
                        "--->\"newestBooks\":[" +
                        "--->--->{" +
                        "--->--->--->\"name\":\"book1\"," +
                        "--->--->--->\"edition\":1" +
                        "--->--->}" +
                        "--->]," +
                        "--->\"newestBookIds\":[" +
                        "--->--->\"" + uuid + "\"" +
                        "--->]" +
                        "}",
                view.toEntity()
        );
    }


    @Test
    public void testEntityToDto() {
        UUID uuid = UUID.randomUUID();
        BookStore entity = BookStoreDraft.$.produce(draft -> {
            draft.setAvgPrice(BigDecimal.ONE);
            draft.addIntoNewestBooks(book -> {
                book.setName("book1");
                book.setEdition(1);
            });
            draft.setNewestBookIds(Collections.singletonList(uuid));
        });

        BookStoreView view = new BookStoreView(entity);
        assertContentEquals(
                "BookStoreView(" +
                        "--->avgPrice=1, " +
                        "--->newestBooks=[" +
                        "--->--->BookStoreView.TargetOf_newestBooks(" +
                        "--->--->--->name=book1, " +
                        "--->--->--->edition=1" +
                        "--->--->)" +
                        "--->], " +
                        "--->newestBookIds=[" +
                        "--->--->" + uuid +
                        "--->]" +
                        ")",
                view
        );
    }
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/dto/BookStoreViewTest.kt
package org.babyfish.jimmer.sql.kt.dto

import org.babyfish.jimmer.sql.kt.common.assertContent
import org.babyfish.jimmer.sql.kt.model.classic.book.addBy
import org.babyfish.jimmer.sql.kt.model.classic.store.BookStore
import org.babyfish.jimmer.sql.kt.model.classic.store.dto.BookStoreView
import kotlin.test.Test

class BookStoreViewTest {
    @Test
    fun testDtoToEntity() {
        val dto = BookStoreView(
            avgPrice = 1.toBigDecimal(),
            newestBooks = listOf(
                BookStoreView.TargetOf_newestBooks(
                    name = "book1",
                    edition = 1
                )
            ),
            newestBookIds = listOf(1)
        )
        assertContent(
            "{" +
                    "--->\"avgPrice\":1," +
                    "--->\"newestBooks\":[" +
                    "--->--->{" +
                    "--->--->--->\"name\":\"book1\"," +
                    "--->--->--->\"edition\":1" +
                    "--->--->}" +
                    "--->]," +
                    "--->\"newestBookIds\":[" +
                    "--->--->" + 1 +
                    "--->]" +
                    "}",
            dto.toEntity()
        )
    }

    @Test
    fun testEntityToDto() {
        val entity = BookStore {
            avgPrice = 1.toBigDecimal()
            newestBooks().addBy {
                name = "book1"
                edition = 1
            }
            newestBookIds = listOf(1)
        }

        val view = BookStoreView(entity)
        assertContent(
            "BookStoreView(" +
                    "--->avgPrice=1, " +
                    "--->newestBooks=[" +
                    "--->--->BookStoreView.TargetOf_newestBooks(" +
                    "--->--->--->name=book1, " +
                    "--->--->--->edition=1" +
                    "--->--->)" +
                    "--->], " +
                    "--->newestBookIds=[" +
                    "--->--->" + 1 +
                    "--->]" +
                    ")",
            view
        )
    }
}